internals/overlord/tlsstate/manager.go

```diff
@@ -0,0 +1,382 @@
+}
+
+// saveIDCert saves the x509 identity certificate, which contains
+// the identity public key.
+func saveIDCert(path string, cert *x509.Certificate) error {
+	certPEMBlock := &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert.Raw,
+	}
+	var pemBuffer bytes.Buffer
+	if err := pem.Encode(&pemBuffer, certPEMBlock); err != nil {
+		return err
+	}
+	pemBytes := pemBuffer.Bytes()
+
+	pemFile, err := os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o600)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		err = errors.Join(err, pemFile.Sync())
+		err = errors.Join(err, pemFile.Close())
+	}()
+
+	if _, err = pemFile.Write(pemBytes); err != nil {
+		return err
+	}
+	return nil
+}
+
+// createIDCert create a public x509 CA certificate, signed by the identity
+// key. This certificate is included in the TLS certificate chain as the
+// non-leaf certificate, allowing the client to pin this certificate during
+// the client server trust exchange (pairing) procedure.
+func createIDCert(signer crypto.Signer) (*x509.Certificate, error) {
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, err
+	}
+	template := &x509.Certificate{
+		SerialNumber:          serialNumber,
+		NotBefore:             systemTime(),
+		NotAfter:              systemTime().Add(idCertValidity),
+		KeyUsage:              x509.KeyUsageCertSign | x509.KeyUsageCRLSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{},
+		BasicConstraintsValid: true,
+		IsCA:                  true,
+		// We can only sign leaf certificates with this.
+		MaxPathLen:     0,
+		MaxPathLenZero: true,
+	}
+
+	// The identity CA cert is self-signed, which allows a client to verify the
+	// incoming version of this with the pinned version.
+	certDER, err := x509.CreateCertificate(rand.Reader, template, template, signer.Public(), signer)
+	if err != nil {
+		return nil, err
+	}
+	// Load a clean X509 certificate from the DER encoded bytes.
+	cert, err := x509.ParseCertificate(certDER)
+	if err != nil {
+		return nil, err
+	}
+	return cert, nil
+}
+
+// isCertActive checks of the current time now is within the start and end time
+// validity period as defined by the certificate.
+func isCertActive(cert *x509.Certificate) bool {
+	now := systemTime()
+	// Note that we shorten the NotAfter timestamp by a tlsCertRenewWindow
+	// duration to avoid a TLS handshake race against the expiry time.
+	if now.Before(cert.NotBefore) || now.After(cert.NotAfter.Add(-tlsCertRenewWindow)) {
+		return false
+	}
+	return true
+}
+
+// isCertDerived checks if the public key in the certificate matches
+// the signer public key. This confirms the certificate matches the
+// identity key.
+func isCertDerived(cert *x509.Certificate, signer crypto.Signer) bool {
+	signerPublic := signer.Public().(ed25519.PublicKey)
+	certPublic := cert.PublicKey.(ed25519.PublicKey)
+	return bytes.Equal(signerPublic, certPublic)
```

Comment:
Should we use [subtle.ConstantTimeCompare](https://pkg.go.dev/crypto/subtle#ConstantTimeCompare) for this (and any similar code)?

Reply:
Sounds good! Done. Resolving.